### PR TITLE
8 animation

### PR DESCRIPTION
**Pull request by Sebastian-Fransson**

### CODE REVIEW

File: Application/src/visuals/Draw.java [hunk 1]
```diff
@@ -164,11 +174,12 @@ public void update() {
      * Renders the message when trying to resize the application.
      */
     void renderMessage() {
+        if(messages.size() == 0) return; // There are no messages in the list.
         if(this.messages.size() > 0) {
-            for (int i = 0; i < messages.size(); i++) {
+            for (int i = 0; i < messages.size(); i++) { //Messages exist and will now be be re-placed.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Draw.java]: Use a iterator instead of a for i loop.

File: Application/src/visuals/Draw.java [hunk 2]
```diff
@@ -38,14 +38,23 @@
...
+     */
     Canvas getCanvas() {
         return canvas;
     }
 
+    /**
+     * Gets the active canvas height.
+     */
     int getHeight() {
         return (int)canvas.getHeight();
```
Comment (kosara-g) [inline Application/src/visuals/Draw.java]: Don't forget to add a return tag

File: Application/src/visuals/Draw.java [hunk 3]
```diff
@@ -38,14 +38,23 @@
...
     }
 
+    /**
+     * Gets the active canvas width
+     */
     int getWidth() {
         return (int)canvas.getWidth();
```
Comment (kosara-g) [inline Application/src/visuals/Draw.java]: Return tag again

File: Application/src/visuals/Draw.java [hunk 4]
```diff
@@ -38,14 +38,23 @@
         canvas = new Canvas(w, h);
     }
 
+    /**
+     * Gets the active canvas.
+     */
     Canvas getCanvas() {
         return canvas;
```
Comment (kosara-g) [inline Application/src/visuals/Draw.java]: Return tag, see below

**Review by IsabelleTornqvist — APPROVED** [review]

File: Application/src/visuals/Draw.java [hunk 5]
```diff
@@ -164,11 +174,12 @@ public void update() {
...
      */
     void renderMessage() {
+        if(messages.size() == 0) return; // There are no messages in the list.
         if(this.messages.size() > 0) {
-            for (int i = 0; i < messages.size(); i++) {
+            for (int i = 0; i < messages.size(); i++) { //Messages exist and will now be be re-placed.
                 Coordinates node1 = diagramClasses.get(messages.get(i).getFromNode()).getCoordinates();
                 Coordinates node2 = diagramClasses.get(messages.get(i).getToNode()).getCoordinates();
-                messages.get(i).changeCoordinates(node1, node2, class_size);
+                messages.get(i).changeCoordinates(node1, node2, class_size); // Changes the coordinates of the messages.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Draw.java]: Long line, Split in to 2 lines.

File: Application/src/visuals/Message.java [hunk 6]
```diff
@@ -43,18 +54,26 @@ public String format() {
...
+           if(keepAnimating == true && node2.getX() > node1.getX()) { // Sending a message.
+               if ((animationBounds += class_size/6) > this.node2.getX() - this.node1.getX())
+                   keepAnimating = false;
+           }
+            //Checks if we are supposed to keep animating, set animationbounds according to how diagramClasses are scaled.
+           else if(keepAnimating == true && node1.getX() > node2.getX()){ // Sending a return message.
+               if((animationBounds -= class_size/6) < (this.node2.getX())  - this.node1.getX())
+                   keepAnimating = false;
+           }
+        } catch (Exception e) {e.printStackTrace();}
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: If you need to catch an update It should be refactored.

File: Application/src/visuals/Message.java [hunk 7]
```diff
@@ -43,18 +54,26 @@ public String format() {
...
+     * Method for animating the message and setting the image size for messages in both directions.
      */
     @Override
     public void update() {
-        // Animate
         try {
-            if ((coolvariable+=10) > this.node2.getX()-this.node1.getX()-2) coolvariable = 0;
-        } catch (Exception e) {}
+            //Checks if we are supposed to keep animating, set animationBounds according to how diagramClasses are scaled.
+           if(keepAnimating == true && node2.getX() > node1.getX()) { // Sending a message.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: if true == true

File: Application/src/visuals/Message.java [hunk 8]
```diff
@@ -43,18 +54,26 @@ public String format() {
...
+               if ((animationBounds += class_size/6) > this.node2.getX() - this.node1.getX())
+                   keepAnimating = false;
+           }
+            //Checks if we are supposed to keep animating, set animationbounds according to how diagramClasses are scaled.
+           else if(keepAnimating == true && node1.getX() > node2.getX()){ // Sending a return message.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: true == true

File: Application/src/visuals/Message.java [hunk 9]
```diff
@@ -64,33 +83,57 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
...
     @Override
     public void render(GraphicsContext gc){
-
         //fromNode Coordinates.
         int x1 = this.node1.getX();
         int y1 = this.node1.getY();
         //toNode Coordinates.
-        int x2 = this.node2.getX();
-        int y2 = this.node2.getY(); // Not used atm.
+        int x2 = this.node2.getX(); //Not used atm.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: Uncomment them if they aren't used?

File: Application/src/visuals/Message.java [hunk 10]
```diff
@@ -64,33 +83,57 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
...
-        gc.fillRect(x1+coolvariable, y1+50, 7, 7);
-        gc.setFill(Color.GREEN);
-        gc.fillRect(x1+coolvariable +2, y1+50 +2, 3, 3);
-        gc.setFill(Color.BLACK);
-
-       // gc.drawImage(image,x1+coolvariable, y1+50, 7, 7 ); // Dragon Image thats going to represent the messages.
 
-       */
+        // Checks if we are supposed to be animating the message.
+        if(keepAnimating==true) {
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: true == true

File: Application/src/visuals/Message.java [hunk 11]
```diff
@@ -64,33 +83,57 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
...
+            //Sets the message text centered above the "dragon".
+            gc.fillText(this.name, x1+animationBounds, y1 + (this.class_size - 2)); // Message description.
+
+            // Checks if up image is supposed to be shown and if we are sending a message and not a return.
+            if (switchImage == true && fromNode < toNode) {
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: true == true

File: Application/src/visuals/Message.java [hunk 12]
```diff
@@ -64,33 +83,57 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
...
+                //sets the dimensions of the dragon according to the current class size.
+                gc.drawImage(dragonMessage, x1 + animationBounds,
+                        y1 + (this.class_size), class_size/1.5, class_size/1.5); //State Wings Up.
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: Set 1.5 to be a variable so you can easier change all of them.

File: Application/src/visuals/Message.java [hunk 13]
```diff
@@ -64,33 +83,57 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
...
+                switchImage = false;
+            }
+
+            // Checks if down image is supposed to be shown and if we are sending a message and not a return.
+            else if (switchImage == false && fromNode < toNode) {
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: !switchImage

File: Application/src/visuals/Message.java [hunk 14]
```diff
@@ -64,33 +83,57 @@ public void changeCoordinates(Coordinates node1, Coordinates node2, int class_si
...
+                //sets the dimensions of the dragon according to the current class size.
+                gc.drawImage(dragonMessage2, x1 + animationBounds,
+                        y1 + (this.class_size), class_size/1.5, class_size/1.5); //State Wings Down.
+                switchImage = true;
+            }
+
+            //Checks if up image is supposed to be shown. if this one is used it is a return message.
+            else if(switchImage==true){
```
Comment (pontuslaestadius) [inline Application/src/visuals/Message.java]: true == true

**Review by pontuslaestadius — CHANGES_REQUESTED** [review]

File: Application/src/visuals/Draw.java (hunk 1 above)
Comment (Sebastian-Fransson) [reply Application/src/visuals/Draw.java]: Not part of this pull-request but i can fix it

**Review by pontuslaestadius — APPROVED** [review]

**Review by kosara-g — APPROVED** [review]

**Pull request merged**